### DISCUSSION TITLE
There was a problem with the PAdES signing code sample.

### DISCUSSION THREAD

**Discussion by projackRK** [MatthiasValvekens/pyHanko#204] [Q&A] [answered]
I have tested this code from Document
![image](https://user-images.githubusercontent.com/39884425/212025957-72dcf550-5bf3-4cfc-ac62-84873d2c9c9f.png)
but i got this error
"pyhanko.sign.general.SigningError: Attempting to sign document with hybrid cross-reference sections while hybrid xrefs are disabled"
My Pdf is version 1.7 and 1.6 
FullCode
from pyhanko.pdf_utils.incremental_writer import IncrementalPdfFileWriter
from pyhanko.sign import signers, timestamps
from pyhanko.sign.fields import SigSeedSubFilter
from pyhanko_certvalidator import ValidationContext

# Load signer key material from PKCS#12 file
# This assumes that any relevant intermediate certs are also included
# in the PKCS#12 file.
signer = signers.SimpleSigner.load_pkcs12(
    pfx_file='test.pfx', passphrase=b'BvYYCHURjG97'
)

# Set up a timestamping client to fetch timestamps tokens
timestamper = timestamps.HTTPTimeStamper(
    url='https://tsa.xxxx.ac.th/tsa/get.aspx'
)

# Settings for PAdES-LTA
signature_meta = signers.PdfSignatureMetadata(
    field_name='Signature', md_algorithm='sha256',
    # Mark the signature as a PAdES signature
    subfilter=SigSeedSubFilter.PADES,
    # We'll also need a validation context
    # to fetch & embed revocation info.
    validation_context=ValidationContext(allow_fetching=True),
    # Embed relevant OCSP responses / CRLs (PAdES-LT)
    embed_validation_info=True,
    # Tell pyHanko to put in an extra DocumentTimeStamp
    # to kick off the PAdES-LTA timestamp chain.
    use_pades_lta=True
)

with open('test.pdf', 'rb') as inf:
    w = IncrementalPdfFileWriter(inf)
    with open('output.pdf', 'wb') as outf:
        signers.sign_pdf(
            w, signature_meta=signature_meta, signer=signer,
            timestamper=timestamper, output=outf
        )

**Comment by MatthiasValvekens** [comment, marked as answer]
Hi @projackRK, this is a standard safety measure to disallow signing hybrid xref files by default, since those files can be parsed in more than one way, leading to ambiguity (and therefore potential security issues). Unfortunately, some software still regularly produces such files (MS Word, for example). If you don't care about this, simply pass `strict=False` to the `IncrementalPdfFileWriter`.

(I should probably reword that error message to make that a bit more obvious)

**Reply by projackRK** [reply]
Sorry, I fixed it as you suggested but I'm still stuck with this Error.
"The signer's certificate could not be validated", InvalidCertificateError('The X.509 certificate provided is not valid for the purpose of non repudiation'
thanks for the answer.

**Reply by MatthiasValvekens** [reply]
You need to relax the key usage policy: https://pyhanko.readthedocs.io/en/latest/api-docs/pyhanko.sign.signers.pdf_signer.html#pyhanko.sign.signers.pdf_signer.PdfSignatureMetadata.signer_key_usage.

**Reply by projackRK** [reply]
I think that might be the reason why I don't have OCSP, but I was able to remove "embed_validation_info=True," and I was able to succeed. Thanks for helping.

**Reply by MatthiasValvekens** [reply]
That's the other way to do it, but of course you lose OCSP response collecting if you go that route :).

**Reply by projackRK** [reply]
Enough to give an example of using signer_key_usage? I still don't understand how to use this param.><"

**Reply by MatthiasValvekens** [reply]
Try passing in

```
signer_key_usage={'digital_signature'}
```

as a parameter to `PdfSignatureMetadata`.

**Reply by projackRK** [reply]
Have this Error 
pyhanko_certvalidator.errors.PathValidationError: The path could not be validated because the end-entity certificate contains the following unsupported critical extension: 1.3.6.1.4.1.311.21.10
From "signer_key_usage={'digital_signature'}"

**Reply by MatthiasValvekens** [reply]
Ah, that's a different issue. Your certificate contains a proprietary Microsoft extension that pyHanko does not support (seems to be something policy-related according to a quick google search). The extension is marked as critical, so pyHanko's validator rejects the certificate (as it is required to do per RFC 5280).

Currently there's no easy way to hook into the validator to implement support for custom extensions, but you can cheat by adding it to the list and replacing this constant: https://github.com/MatthiasValvekens/certvalidator/blob/1c82ddf645a70ecd6871f54c6ef01f6935e28157/pyhanko_certvalidator/validate.py#L920-L937 (don't do that unless you know what you're doing, though). Perhaps disabling validation data embedding is the easiest way forward. You can still fetch OCSP responses manually, if necessary.

**Reply by projackRK** [reply]
Oh, I get it. thank you very much.
